geometric_kernels/spaces/hypercube.py

```diff
@@ -0,0 +1,270 @@
+        """
+        These are certain easy to compute constants.
+        """
+        values = [
+            comb(self.dim, level) * B.ones(float_like(X), *X.shape[:-1], 1)  # [N, 1]
+            for level in range(self.num_levels)
+        ]
+        return B.concat(*values, axis=1)  # [N, L]
+
+    def weighted_outerproduct(
+        self,
+        weights: B.Numeric,
+        X: B.Numeric,
+        X2: Optional[B.Numeric] = None,  # type: ignore
+        **kwargs,
+    ) -> B.Numeric:
+        if X2 is None:
+            X2 = X
+
+        hamming_distances = hamming_distance(X, X2)
+
+        result = B.zeros(B.dtype(weights), X.shape[0], X2.shape[0])  # [N, N2]
+        kravchuk_normalized_j_minus_1, kravchuk_normalized_j_minus_2 = None, None
+        for level in range(self.num_levels):
+            cur_kravchuk_normalized = kravchuk_normalized(
+                self.dim,
+                level,
+                hamming_distances,
+                kravchuk_normalized_j_minus_1,
+                kravchuk_normalized_j_minus_2,
+            )
+            kravchuk_normalized_j_minus_2 = kravchuk_normalized_j_minus_1
+            kravchuk_normalized_j_minus_1 = cur_kravchuk_normalized
+
+            # Instead of multiplying weights by binomial coefficients, we sum their
+            # logs and then exponentiate the result for numerical stability.
+            # Furthermore, we save the computed Kravchuk polynomials for next iterations.
+            result += (
+                B.exp(B.log(weights[level]) + log_binomial(self.dim, level))
+                * cur_kravchuk_normalized
+            )
+
+        return result  # [N, N2]
+
+    def weighted_outerproduct_diag(
+        self, weights: B.Numeric, X: B.Numeric, **kwargs
+    ) -> B.Numeric:
+
+        # Instead of multiplying weights by binomial coefficients, we sum their
+        # logs and then exponentiate the result for numerical stability.
+        result = sum(
+            B.exp(B.log(weights[level]) + log_binomial(self.dim, level))
+            * B.ones(float_like(X), *X.shape[:-1], 1)
+            for level in range(self.num_levels)
+        )  # [N, 1]
+
+        return B.reshape(result, *result.shape[:-1])  # [N,]
+
+    @property
+    def num_eigenfunctions(self) -> int:
+        if self._num_eigenfunctions is None:
+            self._num_eigenfunctions = sum(self.num_eigenfunctions_per_level)
+        return self._num_eigenfunctions
+
+    @property
+    def num_levels(self) -> int:
+        return self._num_levels
+
+    @property
+    def num_eigenfunctions_per_level(self) -> List[int]:
+        return [comb(self.dim, level) for level in range(self.num_levels)]
+
+
+class Hypercube(DiscreteSpectrumSpace):
+    r"""
+    The GeometricKernels space representing the d-dimensional hypercube graph
+    $C^d = \{0, 1\}^d$, the combinatorial space of binary vectors of length $d$.
+
+    The elements of this space are represented by d-dimensional boolean vectors.
```

Comment:
is there a particular reason for using boolean vectors, as opposed to, say, integer vectors?

Reply:
1. To save memory. In the implementation of my dreams, they should be actual bit-arrays. This is much harder than what I did here though, especially if the implementation should be backend-independent.
2. Interpretation. This is a space of binary vectors. Binary ≈ Boolean.

Reply:
Bool requires as much memory as int, so this argument doesn't quite hold. And everyone knows that binary means zeros and ones :)